lib/web/sessions.go

```diff
@@ -122,6 +127,108 @@ func (c *SessionContext) GetClient() (auth.ClientI, error) {
+// returned.
+func (c *SessionContext) GetUserClient(site reversetunnel.RemoteSite) (auth.ClientI, error) {
+	// get the name of the current cluster
+	clt, err := c.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cn, err := clt.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// if we're trying to access the local cluster, pass back the local client.
+	if cn.GetClusterName() == site.GetName() {
+		return clt, nil
+	}
+
+	// look to see if we already have a connection to this cluster
+	remoteClt, ok := c.remoteClt[site.GetName()]
+	if !ok {
+		rClt, err := c.newRemoteClient(site)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// we'll save the remote client in our session context so we don't have to
+		// build a new connection next time. all remote clients will be closed when
+		// the session context is closed.
+		c.remoteClt[site.GetName()] = rClt
+
+		return rClt, nil
+	}
+
+	return remoteClt, nil
+}
+
+// newRemoteClient returns a client to a remote cluster with the role of
+// the logged in user.
+func (c *SessionContext) newRemoteClient(site reversetunnel.RemoteSite) (auth.ClientI, error) {
+	sshDialer := func(network, addr string) (net.Conn, error) {
+		// we tell the remote site we want a connection to the auth server by using
+		// a non-resolvable string @remote-auth-server as the destination.
+		srcAddr := utils.NetAddr{Addr: "web.get-remote-client-with-user-role", AddrNetwork: "tcp"}
+		dstAddr := utils.NetAddr{Addr: reversetunnel.RemoteAuthServer, AddrNetwork: "tcp"}
+
+		// first get a net.Conn (tcp connection) to the remote auth server. no
+		// authentication has occured.
+		netConn, err := site.Dial(&srcAddr, &dstAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// get the principal and authMethods we will use to authenticate on the
+		// remote cluster from our local cluster
+		agent, err := c.GetAgent()
+		if err != nil {
+			err = netConn.Close()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return nil, trace.Wrap(err)
+		}
+		principal, authMethods, err := getUserCredentials(agent)
+		if err != nil {
+			err = netConn.Close()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return nil, trace.Wrap(err)
+		}
+
+		// build a ssh connection to the remote auth server
+		config := &ssh.ClientConfig{
+			User:    principal,
+			Auth:    []ssh.AuthMethod{authMethods},
+			Timeout: defaults.DefaultDialTimeout,
+		}
+		sshConn, sshChans, sshReqs, err := ssh.NewClientConn(netConn, reversetunnel.RemoteAuthServer, config)
+		if err != nil {
+			err = netConn.Close()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return nil, trace.Wrap(err)
+		}
+		client := ssh.NewClient(sshConn, sshChans, sshReqs)
+
+		// dial again to the HTTP server
+		return client.Dial(network, addr)
+	}
+
+	clt, err := auth.NewClient("http://stub:0", sshDialer)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return clt, nil
```

Comment:
probably clt does not support closing directly, but underlying netConn does, so you can return it here with clt explicitly and call AddCloser to close the connection when session ends